cache/remotecache/azblob/exporter.go

```diff
@@ -0,0 +1,214 @@
+			var t time.Time
+			if err := (&t).UnmarshalText([]byte(v)); err != nil {
+				return nil, err
+			}
+			la.CreatedAt = t.UTC()
+		}
+		config.Layers[i].Annotations = la
+	}
+
+	dt, err := json.Marshal(config)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to marshal config")
+	}
+
+	for _, name := range ce.config.Names {
+		if innerError := ce.uploadManifest(ctx, manifestKey(ce.config, name), bytesToReadSeekCloser(dt)); innerError != nil {
+			return nil, errors.Wrapf(innerError, "error writing manifest %s", name)
+		}
+	}
+
+	return nil, nil
+}
+
+func (ce *exporter) Config() remotecache.Config {
+	return remotecache.Config{
+		Compression: compression.New(compression.Default),
+	}
+}
+
+// For uploading manifests, use the Upload API which follows "last writer wins" sematics
+// This is slightly slower than UploadStream call but is safe to call concurrently from multiple threads. Refer to:
+// https://github.com/Azure/azure-sdk-for-go/issues/18490#issuecomment-1170806877
+func (ce *exporter) uploadManifest(ctx context.Context, manifestKey string, reader io.ReadSeekCloser) error {
+	defer reader.Close()
+	blobClient := ce.containerClient.NewBlockBlobClient(manifestKey)
+
+	ctx, cnclFn := context.WithCancelCause(ctx)
+	ctx, _ = context.WithTimeoutCause(ctx, time.Minute*5, errors.WithStack(context.DeadlineExceeded))
+	defer cnclFn(errors.WithStack(context.Canceled))
+
+	_, err := blobClient.Upload(ctx, reader, &blockblob.UploadOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "failed to upload blob %s: %v", manifestKey, err)
+	}
+
+	return nil
+}
+
+// For uploading blobs, use the UploadStream with access conditions which state that only upload if the blob
+// does not already exist. Since blobs are content addressable, this is the right thing to do for blobs and it gives
+// a performance improvement over the Upload API used for uploading manifests.
+func (ce *exporter) uploadBlobIfNotExists(ctx context.Context, blobKey string, reader io.Reader) error {
+	blobClient := ce.containerClient.NewBlockBlobClient(blobKey)
+
+	uploadCtx, cnclFn := context.WithCancelCause(ctx)
+	uploadCtx, _ = context.WithTimeoutCause(uploadCtx, time.Minute*5, errors.WithStack(context.DeadlineExceeded))
+	defer cnclFn(errors.WithStack(context.Canceled))
+
+	// Only upload if the blob doesn't exist
+	_, err := blobClient.UploadStream(uploadCtx, reader, &blockblob.UploadStreamOptions{
+		BlockSize:   IOChunkSize,
+		Concurrency: IOConcurrency,
+		AccessConditions: &blob.AccessConditions{
+			ModifiedAccessConditions: &blob.ModifiedAccessConditions{
+				IfNoneMatch: to.Ptr(azcore.ETagAny),
+			},
+		},
+	})
+
+	if err == nil {
+		return nil
+	}
+
+	if bloberror.HasCode(err, bloberror.BlobAlreadyExists) {
+		return nil
+	}
+
+	return errors.Wrapf(err, "failed to upload blob %s: %v", blobKey, err)
+}
+
+var _ io.ReadSeekCloser = &readSeekCloser{}
+
+type readSeekCloser struct {
+	io.Reader
```

Comment:
Why hide the other functions from interface detection? I would just embed `*bytes.Reader` and `io.Closer`

Reply:
okk let me take a look. wanted to make minimal changes, but open to changing. do you think a  separate follow up PR is fine?